File: March-2023/C++/MinimumCostForTickets.cxx
// Minimum Cost For Tickets
/*
3 Steps
Goal
Choice
Constraint
*/
class Solution {
public:
    int t[366];
    int memoized(vector<int>& days, vector<int>& costs, int& n, int idx) {
          // goal
          if(idx == n)
            return 0; //you can't travel, so no cost
                
        if(t[idx] != -1)
            return t[idx];
        
        // Choice and Constraint 1
        int cost_1 = costs[0] + memoized(days, costs, n, idx+1);
            
        int i = idx;

        // Constraint 2
        while(i < n && days[i] < days[idx]+7) {
            i++;
        }
         // Choice 2
        int cost_7 = costs[1] + memoized(days, costs, n, i);
        
        int j = idx;

        // Constraint 3
        while(j < n && days[j] < days[idx]+30) {
            j++;
        }
        // Choice 3
        int cost_30 = costs[2] + memoized(days, costs, n, j);
        
        return t[idx] = min({cost_1, cost_7, cost_30});
    }
    
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        memset(t, -1, sizeof(t));
        int n = days.size();
        return memoized(days, costs, n, 0);
    }
};

File: March-2023/Java/MinimumCostForTickets.java
// Minimum Cost For Tickets
/*
3 Steps
Goal
Choice
Constraint
*/

class Solution {
    int[] arr = new int[365];
    public int solve(int[] days, int[] cost, int n, int index){
        
        //goal
        if(index>=n){
            return 0;
        }
        
        // memoization
        if(arr[index]!=-1){
            return arr[index];
        }
        
        // choice  and constraint 1
        int cost_1 = cost[0]+solve(days,cost,n,index+1);
        
        int i = index;
        
        // contraint 2 
        while(i<n && days[i]<days[index]+7){
            i++;
        }
        // choice 2
        int cost_2 = cost[1]+solve(days,cost,n,i);
        
        int j = index;
        
        // constraint 3
        while(j<n && days[j]<days[index]+30){
            j++;
        }
        
        // choice 3
        int cost_3 = cost[2]+solve(days,cost,n,j);        
        return arr[index]=Math.min(cost_1,Math.min(cost_2,cost_3));
               
    }
    public int mincostTickets(int[] days, int[] costs) {        
        int index = 0;       
         Arrays.fill(arr,-1);
        int n = days.length;
        return solve(days,costs,n,index);
       
    }
}
